### DISCUSSION TITLE
Plan for supporting data sources

### DISCUSSION THREAD

**Discussion by wangxiaoying** [sfu-db/connector-x#61] [Ideas]
We plan to add more RDBMS to connector-x as data sources. Here is a list that we are planning to support next:

Stage 1 (leading open source sources)
- MariaDB (MySQL-derivative)
- MySQL

Stage 2 (popular commercial sources)
- Oracle
- Microsoft SQL Server
- Microsoft Azure SQL Database

Stage 3 (popular cloud-based PaaS that can leverage completed work-to-date)
- Amazon Redshift (Postgres-derivative)
- Amazon Aurora (MySQL-derivative)
- Google BigQuery

Stage 4 (popular open source sources)
- Hive
- Impala
- Spark SQL
- Presto / Trino

Backlog:
- SQLite
- DB2
- Snowflake

Feel free to reply here if you think we should add any other RDBMS / data sources (e.g. S3), or share your opinion about the priority of implementation!

**Comment by pablojim** [comment]
Clickhouse support would be great!

**Reply by pcoccoli** [reply]
I agree on Clickhouse!

**Reply by overcoil** [reply]
The good news is Clickhouse supports the MySQL wire protocol and development for MySQL is under way already.

**Reply by dovahcrow** [reply]
Hi @pcoccoli, since Clickhouse uses the MySQL protocol, you can use the MySQL support in our 0.2.0a2 release to download data from Clickhouse.

**Reply by dovahcrow** [reply]
We currently support these data types for MySQL: Int, BigInt, Double, Datetime, Date, Time, Decimal, Char, VarChar.

**Comment by wolfgangcolsman** [comment]
Trino support would be great as well!

**Reply by andreapiso** [reply]
Trino itself only sends you data through REST API so probably not much advantage in using connectorx? Since the bottleneck is o  their side.

The trino jdbc and odbc call the REST API and convert from json...

**Comment by pcoccoli** [comment]
I think SQLite should be prioritized higher since it's so common and easy to use.

**Reply by dovahcrow** [reply]
Hi @pcoccoli  we are pleased to inform you that we just released the connectorx `0.2.0a2` version, which includes MySQL and SQLite support. You can try it using `pip install -U connectorx==0.2.0a2`.

**Reply by pd2871** [reply]
@dovahcrow can you write the code sample to connect the sqlite3? I have a local .db file and I want to connect it with connectorx.

**Reply by wangxiaoying** [reply]
Please check out our doc [here](https://sfu-db.github.io/connector-x/databases/sqlite.html).

**Comment by mrtztg** [comment]
Great. I'm waiting for MySQL support.

**Reply by mrtztg** [reply]
When MySQL will be supported? I'm waiting for it to release my app.

**Reply by dovahcrow** [reply]
Sorry for the waiting. I'm currently reviewing the MySQL support PR at https://github.com/sfu-db/connector-x/pull/89. Hopefully we will be able to release an alpha version next week, including the support for sqlite and mysql.

**Reply by dovahcrow** [reply]
Hi @mrtztg we are pleased to inform you that we just released the connectorx `0.2.0a2` version, which includes MySQL and SQLite support. You can try it using `pip install -U connectorx==0.2.0a2`.

A simple benchmark on the TPC-H lineitem table (8G) shows we are 36x faster than pandas + PyMySql (shown in the gif), 15x faster than pandas+mysqlconnector (the MySQL connector provided by MySQL) and 12x faster than pandas + mysqlclient (the fastest python MySQL connector).

![tpch-mysql](https://user-images.githubusercontent.com/998606/125568464-ad83401b-b693-4f62-a7cf-55d94b6872ba.gif)

**Reply by mrtztg** [reply]
Thanks @dovahcrow ❤️❤️❤️. I'm going to try it.

**Reply by mrtztg** [reply]
It runs ~40% faster without using partitioning.
This is good. My query is complex. I will try to use partitioning and give the result here.

**Reply by jokyv** [reply]
I tried with connectorx==0.2.0a2 but I got "PanicException: not implemented: MYSQL_TYPE_TIMESTAMP"

**Reply by dovahcrow** [reply]
@jokyv Yep we haven't added the timestamp type to the code. Will do that in the next release.

**Comment by ericjohn** [comment]
KDB+/Q support would be nice to have!

**Comment by JazzFan** [comment]
I would like to ask for BigQuery to have a higher priority

**Comment by cawoodjm** [comment]
When do you think Oracle and Microsoft SQL Server will be ready for testing or generally available?

**Reply by wangxiaoying** [reply]
Hi @cawoodjm, we just released the connectorx-0.2.1a1 which includes both [Oracle](https://github.com/sfu-db/connector-x/blob/main/Types.md#oracle) and [Microsoft SQL Server](https://github.com/sfu-db/connector-x/blob/main/Types.md#sqlserver). You can try it using `pip install -U connectorx==0.2.1a1`. The benchmark result for [Oracle](https://github.com/sfu-db/connector-x/blob/main/Benchmark.md#oracle-dbr54xlarge-rds) and [Microsoft SQL Server](https://github.com/sfu-db/connector-x/blob/main/Benchmark.md#mssql-r54xlarge-docker-in-another-ec2-instance) are also uploaded respectively.

If all tests go well, we plan to release the 0.2.1 version in early October. Please feel free to have a try and let us know if you found any issue.

**Comment by abel-diaz** [comment]
I am unable to connect to Oracle. I am using Connector X v. 0.2.1a1
I have the following creds
host
port
servicename
username
password

According to the connectorx instructions i need to have
(postgres|postgressql|mysql|mssql|sqlite)://username:password@addr:port/dbname

Using my above creds, I have
cstr = 'oracle://username:password@host:port/servicename'
query ' 'select * from table'

When I run
df = cx.read_sql(cstr,query)

I receive RuntimeError: timed out waiting for connection: OCI Error: ORA-01017: invalid username/password; logon denied
I am able to connect with sqlalchemy using my creds

How do I connect conncetorx with Oracle?
Thank you,
Abel

**Reply by cawoodjm** [reply]
I am experiencing similar issues, I am unable connect to an oracle environment no matter how I configure the connection string.

**Reply by wangxiaoying** [reply]
Hi @cawoodjm , this issue has been solved here: https://github.com/sfu-db/connector-x/discussions/130. Not sure whether our newest alpha version (now is 0.2.2-alpha.1) can help.

**Comment by farhat-nawaz** [comment]
Is `daterange` type for postgres not supported yet? I'm getting a `PanicException: not implemented: daterange` error.

**Reply by wangxiaoying** [reply]
Hi @farhat-nawaz , we do not support `daterange` type in postgres right now since we didn't find the parsing of this type implemented in the underlying [driver](https://docs.rs/postgres/0.19.2/postgres/types/trait.FromSql.html). All the types we support are listed [here](https://github.com/sfu-db/connector-x/blob/main/Types.md#postgres-redshift).

**Comment by matthieugross-bib** [comment]
Hi, I'm trying to connect to a Microsoft Azure SQL Database.  It works fine using pymssql:

```
conn = pymssql.connect(user='a',
                                        password='b',
                                        database='c',
                                        host='d.database.windows.net',
                                        port=e)
query = "SELECT * FROM INFORMATION_SCHEMA.TABLES order by TABLE_SCHEMA,TABLE_NAME"
schema = pd.read_sql(query,conn)
schema.head()
```


But I get a "Timed out in bb8" runtime error using the following:
```
import connectorx as cx
conn_url = "mssql://a:b@d.database.windows.net:e/c"
query = "SELECT * FROM INFORMATION_SCHEMA.TABLES order by TABLE_SCHEMA,TABLE_NAME"
df = cx.read_sql(conn_url, query)

```
See:
```
.../site-packages/connectorx/__init__.py in read_sql(conn, query, return_type, protocol, partition_on, partition_range, partition_num, index_col)
    121             queries=queries,
    122             protocol=protocol,
--> 123             partition_query=partition_query,
    124         )
    125         df = reconstruct_pandas(result)

RuntimeError: Timed out in bb8
```

Currently using version '0.2.3'.

Any idea?
Thanks 
Matt

**Reply by wangxiaoying** [reply]
Hi, it seems like we currently have some issue with Azure database. I will redirect it to this discussion: https://github.com/sfu-db/connector-x/discussions/181#discussioncomment-2229577.

**Reply by anybodysguest** [reply]
I am actually getting the same error.  I am trying to connection to a MS SQL Server in AWS EC2.  I used the example shown here:[
](https://github.com/sfu-db/connector-x/blob/main/docs/databases/mssql.md)

There is a same trusted connection that is shown that looks like this:  mssql://host:port/db?encrypt=true&trusted_connection=true

So, in my case I replace the values in the same above with my own.  For example:
Server/Host = MyMsSQLServer
Port=1433
Database = MyDatabase

Now the connection string looks like this:
mssql://MyMsSQLServer:1433/MyDatabase?encrypt=true&trusted_connection=true

When I attempt to connect I am getting the "Timed out in bb8" that @wangxiaoying  pointed out.

I'd like to use ConnectionX because I am trying our Polars.  Polars, apparently using ConnectionX under the covers.

Any ideas?

**Reply by wangxiaoying** [reply]
Hi @anybodysguest , does this comment (of enabling TCP/IP) help? https://github.com/sfu-db/connector-x/discussions/181#discussioncomment-2776956

**Reply by anybodysguest** [reply]
I'm pretty sure all of our MS SQL Servers have TCP/IP enabled.

**Comment by wKollendorf** [comment]
Hi,

are there any plans for integrating impala?

Thanks

**Reply by wangxiaoying** [reply]
Hi @wKollendorf , we are currently working on supporting ODBC #215 , which could be used to access impala when it's ready.

**Comment by grantmwilliams** [comment]
I think integrating with S3 or Athena would be pretty useful

**Reply by overcoil** [reply]
Are Trino (formerly PrestoSQL) or PrestoDB of use/interest? They are upstream of Athena but different.

**Reply by alec-heif** [reply]
trino would be great!

**Comment by ameetnsharma** [comment]
Hi. I'm getting this error in AWS:

pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }

So I'm in a loop that's calling

df = cx.read_sql(f"mysql://{db_username}:{db_password}@{db_host}:3306/{db_name}", query)

where query is changing to cover different date ranges.

The loop works for 336 times then the exception happens. On the 337th time it gives this exception. So it seems to be unrelated to the particular date range I'm pulling from... if I start the loop at a different starting date... it still runs exactly 336 times then crashes on the 337th time.

Appreciate any help. Thanks.

EDIT: Very similar issue on another resource in case it's of any help:

https://github.com/huggingface/tokenizers/issues/444

I was able to fix the issue on my end by pulling more data per loop and reducing the number of loops. Still, might be something to look at.

**Reply by wangxiaoying** [reply]
Hi @ameetnsharma,

I tried to run a query in a loop 400 times on an AWS instance to fetch data from mysql, but it works fine.

I think this error is more similar to this issue: https://github.com/openethereum/parity-ethereum/issues/10009 where might be caused by not enough resource (RAM, reaching to thread limit, etc) when tokio spawn new thread. Can you check whether your instance's status when you see the error?

**Comment by ameetnsharma** [comment]
Hi Mr. Wang. I was able to fix the problem. It was likely something wrong with my instance. But I'll keep a lookout if it happens again. Thanks. Excellent tool! Best way to read from sql into pandas.

-Ameet

________________________________

**Comment by sfc-gh-paubrey** [comment]
Snowflake please!

**Reply by wangxiaoying** [reply]
The odbc driver support is currently under development. connectorx should be able to connect to snowflake through odbc when it's done!

**Reply by sfc-gh-paubrey** [reply]
Thank you! If you need me to help in anyway then happy to test with you.

**Reply by sfc-gh-paubrey** [reply]
@wangxiaoying do you have any timelines on when ODBC may be available?

**Reply by wangxiaoying** [reply]
Hi @peteraubreysf , we don't have a precise timeline for now. But it should be done by next month.

**Reply by sfc-gh-paubrey** [reply]
Thank you!

**Reply by sfc-gh-paubrey** [reply]
Hi @wangxiaoying how is it going? Hope you are well. I have a client who would love to see Snowflake ConnectorX working with Snowflake. Any update on your support for Snowflake? Anyway I can help at all?

**Reply by wangxiaoying** [reply]
Sorry that the [ODBC client driver](https://docs.rs/odbc-api/latest/odbc_api/) is quite different from other databases that we previously supported, and therefore we have to delay this support. @Verus121 is currently working on this. He has some initial code on this [branch](https://github.com/sfu-db/connector-x/tree/odbc) with a basic framework but the functionalities are not supported yet. Sorry for the inconvenience and please feel free to contribute (check out this [file](https://github.com/sfu-db/connector-x/blob/odbc/connectorx/src/sources/odbc/mod.rs) for supporting the functionalities on ODBC) : )

**Reply by sfc-gh-paubrey** [reply]
Thanks @wangxiaoying, has any more progress been made do you know? Apologies, not being a developer myself I can't help with the coding work!

**Reply by verramilli** [reply]
hi @wangxiaoying : Any update on snowflake? ConnectorX works very well for mysql but would like to use it for Snowflake as well. Thanks.

**Reply by zykezero** [reply]
amazing news

**Reply by GeraudK** [reply]
@wangxiaoying would a valid option to have a fallback when the client driver isn't supported by ConnectorX to use a `SQLAlchemy connectable` like [pandas.read_sql](https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html) does?

**Reply by MarvaZy** [reply]
Do you have an estimation when the support for snowflake will be available though? I see the branch you mentioned in the comments is not done yet

**Reply by colelandolt** [reply]
Curious as to whether support for snowflake is still in the works and if so, is there is a timeline? I see that the branch mentioned above hasn't been contributed to in almost one year. Thanks for all of your work on this thus far!

**Reply by connor-elliott** [reply]
@wangxiaoying is Snowflake support still planned/being worked on? Thank you

**Reply by abstractqqq** [reply]
Snowflake support would be highly appreciated!

**Comment by rferraton** [comment]
SAP HANA would be great !

**Comment by chitralverma** [comment]
Apache Hive would be quite good. Is there a issue/ PR or any ongoing work that I can possibly contribute to ?

**Reply by wangxiaoying** [reply]
Hi @chitralverma , we are currently developing the support for ODBC. Which might be used to support connecting to Hive : )

**Reply by andreapiso** [reply]
ODBC and hive are two different things tho! ODBC uses HiveQL engine which is quite slow. What would be amazing to see is native integration with the HMS (the way spark does it for example)

**Comment by guillehampp** [comment]
Hi, 
I tried to query a table with "geometry" but i get this error:
pyo3_runtime.PanicException: not implemented: geometry

**Comment by earlev4** [comment]
Hi @wangxiaoying! Thank you so much for your efforts on ConnectorX. The benchmarks look very promising. I hope in the future, there is consideration for adding [DuckDB](https://duckdb.org/) to the list of data sources. I would be very excited to see a benchmark performance measurement against the other Python solutions. Not sure if helpful, but there is already a [duckdb crate](https://crates.io/crates/duckdb) and the r2d2 feature can be used from [duckdb-rs](https://github.com/wangfenjin/duckdb-rs) according to the [r2d2 crate](https://crates.io/crates/r2d2). Thanks again for the wonderful project!

**Reply by wangxiaoying** [reply]
Hi @earlev4 , thanks a lot for the great suggestion! We do think DuckDB is an important data source. But it seems like DuckDB's developers already provided a solution for acquiring SQL result in the format of pandas in [this blog post](https://duckdb.org/2021/05/14/sql-on-pandas.html), so we are not sure how much they already do. We will do some investigations to see whether it is good enough or we can further accelerate in connectorx : )

**Reply by earlev4** [reply]
Hi @wangxiaoying! Thank you so much! I appreciate it!

**Comment by DineshS-nd** [comment]
Hi my database does not have any evenly  distributed column for partition. Is there any way to partition based on time stamp ?

**Reply by wangxiaoying** [reply]
Hi @DineshS-nd , we do plan to support partition on timestamp in the future (#313). But currently, a workaround is to partition the query on timestamp column manually like:

```
import connectorx as cx

conn = "postgresql://username:password@server:port/database"
queries = ["SELECT * FROM table WHERE timestamp <= '2020-01-01'", "SELECT * FROM table WHERE timestamp > '2020-01-01'"]

cx.read_sql(conn, queries)
```

**Comment by wangxiaoying** [comment]
tdengine #341

**Comment by overcoil** [comment]
Hi,

When you say Trino, are you referring only to OSS Trino? Is there any need
for Athena? Or the Starburst variants?

George

On Mon, Oct 17, 2022 at 1:00 AM Peter Aubrey ***@***.***>
wrote:

> Thanks @wangxiaoying <https://github.com/wangxiaoying>, has any more
> progress been made do you know? Apologies, not being a developer myself I
> can't help with the coding work!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sfu-db/connector-x/discussions/61#discussioncomment-3893919>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMP7OQCY3OVT5XJTGQYCLTWDUBTRANCNFSM422E6IZQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
--
"Not everything that can be counted counts, and not everything that counts
can be counted." Albert Einstein

**Reply by romanzdk** [reply]
For us OSS Trino is enough :D

**Comment by fabricebaranski** [comment]
Druid would be cool

**Comment by wnsgudcnt** [comment]
I hope Teradata Connection....!!

**Comment by senordeveloper** [comment]
line 224, in read_sql
    result = _read_sql(
pyo3_runtime.PanicException: not implemented: interval

Any ideas to this error? Used the latest version.

**Reply by senordeveloper** [reply]
It's regarding the standard postgres connection

**Comment by JaimeArboleda** [comment]
Hello! I don't see Denodo mentioned anywhere, is it supported? I know it's a gateway, but we have it in our installation and it's our only way to access underlying datasets.

**Comment by karrtikiyer** [comment]
Support AWS Athena would be great!

**Comment by RxL-Mudit-Singh** [comment]
Reading data from Oracle tables and storing it in a spark data frame will  be of  great help

**Comment by satyakommula96** [comment]
Apache Calcite Avatica support would be great!

**Comment by daincrawford** [comment]
Seems like you could make a lot of people happy if they had ODBC read-only support for existing viable DSN's. And then work on possible to_sql later. 
Keep up the great work. Thank you @wangxiaoying.

**Comment by romanzdk** [comment]
Hello, any update on Trino connector?

**Comment by agalvanz** [comment]
Is it possible to add Vertica support?

**Comment by SantiScandorla** [comment]
I would like to connect SQL Server Management Studio with Python using polars and connectorx function:

I wrote this chunk of code

```
import connectorx as cx 
import urllib.parse

conn = 'mssql://sa:{}@{}:5432/Planning_insight?encrypt=true&trusted_connection=true'.format(
    urllib.parse.quote_plus('MYPASSWORD'),
    urllib.parse.quote_plus('MYSERVER\SQLEXPRESS')
    )                                                                        
str_query = f"SELECT * FROM raw.calendar"

cx.read_sql(conn,str_query)
```
but after 30 second, it returned this error

RuntimeError Traceback (most recent call last) Cell In[1], line 10 4 conn = 'mssql://sa:{}@{}:5432/Planning_insight?encrypt=true&trusted_connection=true'.format( 5 urllib.parse.quote_plus('MYPASSWORD'), 6 urllib.parse.quote_plus('MYSERVER\SQLEXPRESS') 7 )
8 str_query = f"SELECT * FROM raw.calendar" ---> 10 cx.read_sql(conn,str_query)

File c:\Users\lorenzsc\AppData\Local\miniconda3\envs\padp2\lib\site-packages\connectorx_init_.py:224, in read_sql(conn, query, return_type, protocol, partition_on, partition_range, partition_num, index_col) 221 except ModuleNotFoundError: 222 raise ValueError("You need to install pandas first") --> 224 result = _read_sql( 225 conn, 226 "pandas", 227 queries=queries, 228 protocol=protocol, 229 partition_query=partition_query, 230 ) 231 df = reconstruct_pandas(result) 233 if index_col is not None:

RuntimeError: Timed out in bb8

Can anyone help me?

**Reply by pangjunrong** [reply]
Hey @SantiScandorla, SSL was previously unsupported for MSSQL which resulted in a generic RuntimeError when encrypt=true&trusted_connection=true was specified. As of 0.4.0, you should be able to connect with no issues now by adding `trust_server_certificate=true.` together with the other parameters.

**Comment by mvpalheta** [comment]
I suggest support for Netezza

**Comment by trichie** [comment]
Is there any chance for DB2 support to be added in the near future?

**Comment by nicolaslattuada** [comment]
DuckDb and azure data explorer would be nice too !

**Reply by bsplosion** [reply]
Seconding DuckDB.  We have a need to frequently migrate data from Postgres to DuckDB and I suspect ConnectorX would have better performance for that use case with the federated database feature than the existing Postgres extension on DuckDB has today.

**Comment by j0bekt01** [comment]
Are there any plans to also support writing to DBs?

**Comment by hnbn313** [comment]
adding teradata to supported dbms would be great !!!

**Comment by vebalaji** [comment]
Does this currently support connecting to and retrieving data from Amazon Aurora PostgreSQL tables?

**Comment by TsinlingLi** [comment]
Can the native protocol of clickhouse be supported? It is not very practical to obtain data from clickhouse using the MySQL protocol because many places control port numbers, and in most cases, only the native port numbers are opened to the external network

**Comment by mhvsa** [comment]
Hi!

I appreciate your work, fantastic job.

Any plans supporting oracle wallet as auth in oracle connection? If it’s supported, then some notion in the documentation would be helpful.

Thanks!

**Reply by pangjunrong** [reply]
Hey @mhvsa, assuming you have your tnsnames.ora setup and ready to go with your wallet, you can use the alias connection method. This is found in the documentation under [Oracle TNS Names (Alias) Connection](https://sfu-db.github.io/connector-x/databases/oracle.html).

You can find more information about [configuring your wallet with TNS here](https://docs.oracle.com/middleware/1213/wls/JDBCA/oraclewallet.htm#JDBCA604).

**Comment by Feromond** [comment]
Definitely am interested to see if there could be some support added for databricks sql, or if there is a similar protocol that exists already that could work for connecting and making queries to data bricks.

**Comment by willshiao** [comment]
Hi, I would love it if you could add ScyllaDB support. It's great for low-latency operations and is compatible with Apache Cassandra so support for both would probably come together. It also has a SQL-like query language (CQL), so the UX wouldn't have to change much.

**Comment by yufanlu** [comment]
It would be wonderful to support Vertica

**Comment by sector119** [comment]
It would be great to have support for reading DBF files as a data source

**Comment by Zahlii** [comment]
Probably very niche but relevant for my use case, but SAP HANA support would open up a lot of possibilities in the SAP / business ecosystem for speed improvements

**Comment by panchoferrari** [comment]
It would be be great to have support for Databricks SQL or databricks delta lake tables
